src/url-params.ts

```diff
@@ -175,11 +175,24 @@ export class AnalyticsUrlParams {
 	/**
 	 * @return full query parameter string that can be appended to URLs
 	 */
-	getQueryString(destinationUrl?: URL, currentUrl?: URL): string {
+	getQueryString(destinationUrl?: URL | string, currentUrl?: URL): string {
+		// we first check if destionationUrl is a relative URL string. If it is, we exit.
+		const relativeRegex = /^(?!www\.|(?:http|ftp)s?:\/\/|[A-Za-z]:\\|\/\/).*/g;
+		if (
+			typeof destinationUrl === 'string' &&
+			destinationUrl.match(relativeRegex)
+		) {
+			return '';
+		}
+
+		// If its not a relative URL string, we make sure if its a string or an URL. It its a string, we convert it to an URL object.
+		if (typeof destinationUrl === 'string') {
```

Comment:
You can combine this `if` statement with the one above to eliminate the double check for `string`. The code and comments will also read better.
```js
if (typeof destinationUrl === 'string') {
    if (destinationUrl.match(relativeRegex)) {
        return '';
    } else {
        ....
}

Reply:
@gelbal Lol I was going to do it that way first, but for some reason I remembered that once I saw the if statements one below the other and that that was the usual practice or something like that and went the other way. Sure I'll change that.

Reply:
@laziob oh I see you made a major rework. Actually I liked your previous implementation as it was cleaner (once you combine the new `if` statements). So the code handles the `string` scenario first and the proceeds to handling the `URL` scenario. Can you please bring it back? : )

Reply:
Petros (ex-CTO) made a pretty good presentation about such design actually. I highly recommend. 
https://docs.google.com/presentation/d/1hFV5YBL4FflfN7OIJr6eZrUV4Pk158wOXWzea1ANTnM/edit#slide=id.p
https://drive.google.com/file/d/1xtHev7HUwbASdasf8cWfG9O2oHdlXX43/view

Reply:
@gelbal It still handles the string scenario first. The code above is for the other parameter that stays that way regardless of the `destinationUrl`. The reason I changed it was because with the previous design, I had to repeat much more code, once for each if and even then couldnt make it work properly. In fact I dont even like this design, since it still repeats code. But it repeated less than the previous one. I still also dont believe we should be handling relative URLs through this method. It kind of undermines the use of TypeScript and gives place to the bad use of the method in `resin-site`, since I still think they shoudnt be forcing this method in every link "just in case". Performance wise it doesnt make sense to me. `resin-site` should have a better way to identify outbound links from inbound links, and call this method only on outbound links. 

Having said that, I could try go back to the previous design, but I could make it work with the previous design. This was so far the only way I achieved what I wanted lol.

Reply:
> It kind of undermines the use of TypeScript

I know what you mean. I checked and apparently TypeScript does not yet support string types based on regex. See the issue: https://github.com/microsoft/TypeScript/issues/41160
Regex-validated string types would be great but alas we don't have them now.

As for `normalizing if`: Your previous solution was cleaner as it first handles the scenario of receiving `string` parameter and then once it's handled, the rest of the logic stays the same for `URL` parameter scenario.

Reply:
Is there a simpler way for UI / React developers to signal relative URLs on their end? 

I understand the frontend folks' point of view as they want a convenient method that handles all scenarios to generate URL params. I don't think it's bad or lazy on their side to pass any URLs to this function. At the end `analytics-client` is a module and you could think of it as an extension of the tools that frontend folks use as part of their own codebase. So it's totally fine for the module to be responsible for such validations.

Reply:
By no means Im saying that its because of lazyness or unwillingess to do it. Is just that adding the URL params is something that has to be done in really few cases relative to the amount of links a web component might have. So calling if for all links seems backwards. I dont know much about performance to be honest, but is like building a kind of sorter, when actually the cases that you need to find positive are really few and can be pinpointed. For example the way it is now, we would also be appending Url Params to even external links like twitter.com, github.com, youtube.com, and many sites that dont know what does params even are, which it might even break down links since we dont control what happens on the other side when URL params are present. So to solve for that this method should also handle a way to identify just balena domains. For that we would then need another regex (which are not bulletproof) or a list somewhere that would need maintenance. It feels like building a big machine for doing a small task.

Reply:
I hear you Ezequiel. I agree that it'd be neater if the module user has a simple way to detect relative URLs (and also external 3rd party sites) on their app code, and in such cases not call this `getQueryString` method . Doesn't seem like so, right? 

Yeah, an explicit list of domains to keep tracking would be neat. I know we discussed this briefly before. Perhaps we could revisit this topic, given the increasing complexity when we don't have such an explicit list. It might not be too much to ask from component maintainers to provide `analytics-client` a list of potential domains that a user could navigate to.

Reply:
To put an example, `Hub` handles it like this:

`	deployUrl = getExternalLinkWithTracking(
		'https://dashboard.balena-cloud.com/deploy?repoUrl=${repoUrl}')`

So it explicitly uses the method on the only domain where its needed. For example marketing site only needs to use it on the Login and Signup buttons, which exist only in 2 places, the header (for most of the resin-site pages) , like [here](https://github.com/balena-io/resin-site/blob/90a8eaf39cc33298e558a744171b8ccd58e38d5c/src/components/Header/Header.js#L246) and the box for on the landing page. 

So its not that they need to detect relative urls in a simple way. Its that they have to detect URLs to another balena domain, not even including subdomains. 

If using a list is the best way to do it I dont really know, but if its needed to be raised for further discussion Im up for that. Whats the proper way and place to raise it?  ALthough It just seems too much for the task at hand tbh.

Reply:
@gelbal we need to work on this soon since its the only thing preventing marketing sites to update to the last analytics client. Its still using 0.12 so its not stiching sessions nor using the last referrer updates.

Reply:
@iamsolankiamit @thgreasi @JSReds if you could take a look into the discussion too so we can decide on how we advance it would be really helpful and appreciated! thanks!

Reply:
Hey @laziob cc @thgreasi @JSReds @iamsolankiamit, how about introducing an optional parameter (in the `analytics-client` constructor) for a list of domains to track?

We could retrieve this list either directly from the client that calls `analytics-client` or indirectly from the `projectName` configuration stored in the `analytics-backend`. At first it'll have the following content:
```
*.balena.io
*.balena-cloud.com
```
We will probably end up updating this list seldom so it's not a big maintenance issue to store it on the `analytics-backend` side.

Having said these, let's still go ahead with working on this PR to add support for relative URL detection.

Reply:
Having said these, let's still go ahead with working on this PR to add support for relative URL detection.